## metaMLP
MetaMLP is a fast approach to annotate metagenomic datasets agains a protein reference database. It compares sequence reads against a protein database, similar to what blastx does. However, instead of performing sequence alignment, it uses word embeddings to represent kmers as vectors. MetaMLP is built from the FastText library for text classification in combination with the seqan library to perform fast I/O opperations and to perform the six reading frame translation. To index protein sequences (reference database) we use the reduced alphabet proposed in RAPSearch. metaMLP is several orders of magnituf faster than blastx and diamond.

Performance of MetaMLP is comparable to diamond blastx with an identity cutoff of 80%, but, its around 40 times faster.

## Installation

        git clone https://github.com/gaarangoa/ARGfast.git
        cd ARGfast
        make clean && make

### Usage
#### metaMLP
        usage: predX <command> <args>

        The commands supported by predX are:

        index              Index the reference sequences
        quant              Run quantification algorithm [fasttext model]

#### Index
        usage: predX index <args>

        The commands supported by predX index are:

        -input        Protein reference database
        -output       Output index
        -kmer         k-mer size in aminoacids [default 11]
        -labp         Label index position in the FASTA header (default 4: >xx|xx|xx|label|xx)
        -NoReduced    Dissable the reduced alphabet and use all 20 Amino Acids

#### Annotation
        usage: predX quant <args>

        The commands supported by predX quant are:

        -model        Trained model - generated by predX index
        -input        FASTA file with short sequence reads
        -output       output file to write the processed reads
        -proc         Number of threads to use [default 8]
        -kmer         k-mer size [default 11 - same kmer used in predX index]
        -seed         seed size [default 11 - amino acids]
        -mink         minimum number of kmers that each read has to contain [default 5]
        -NoReduced    Enable it if index is built with the -NoReduced option

To create the index of the Antibiotic Resistance profiling, you can use our developed database [1] (please check the publication).

### Index protein reference file
To index a protein reference database you first need to add the gene labels of the database to the header of the fasta file. For instance, just add the label to one of the fields in the header separated by the character "|" as follows >gene_id|field1|label|field3. For this particular example the index where the label is placed is 3.

        ./metaMLP index \
            -input /path/to/reference_protein_database.fasta \
            -output /path/to/output/data/prefix \
            -labp 3 \
            -kmer 11 \
            -dim 64 \
            -epoch 50

Where -labp corresponds to the index where the label is placed within the fasta header and prefix is the database index name. metaMLP generates two files {prefix}.bin and {prefix}.kh.


### Perform Functional Profiling
To perform the functional profiling you need to use the quant command. The input file has to be a fasta file, it does not support fastq files, to get a fasta file you can use another tool such as vsearch to merge paired end reads. The -model depicts the model created by ./predX index.

        ./metaMLP quant \
            -model model \
            -input input_reads.fasta \
            -output output_file \
            -kmer 11 \
            -tries 2

